include/aws/common/common.h

```diff
@@ -52,6 +54,14 @@ AWS_COMMON_API void *aws_mem_acquire(struct aws_allocator *allocator, size_t siz
  */
 AWS_COMMON_API void aws_mem_release(struct aws_allocator *allocator, void *ptr);
 
+/*
+ * Attempts to adjust the size of the pointed-to memory buffer from oldsize to
+ * newsize. The pointer (*ptr) may be changed if the memory needs to be reallocated.
+ *
+ * If reallocation fails, *ptr is unchanged, and this method raises an AWS_ERROR_OOM error.
+ */
+AWS_COMMON_API int aws_mem_realloc(struct aws_allocator *allocator, void **ptr, size_t oldsize, size_t newsize);
```

Comment:
could we add some documentation for future generations as to why this api appears inconsistent, but really isn't? This will avoid someone doing a costly, and wrong refactor 2 years from now.